Field split decomposition bug fixes

lindsayad opened the pull request:
- For `navier_stokes` field splits: Back when this embedding code was written we didn't support vector variables in MOOSE. However, now we do so we have to make sure we get all the dof components of the vector variables
- For `contact` field splits: we were over-providing dofs both for the contact and interior portions of the splits. 
- With the changes in this PR we now satisfy error checking that the the sum of the splits are equal to the system matrix size both for `navier_stokes` and `contact`

Refs #22359

lindsayad committed:
Print IS size

lindsayad committed:
Error if splits don't add to local system matrix size

lindsayad committed:
Loop over all dof components for a node for a given var

Back when this embedding code was written we didn't support vector
variables in MOOSE. However, now we do so we have to make sure
we get all the dof components of the vector variables

Refs #22359

lindsayad committed:
Only perform size checking in devel/dbg modes

Looking at the code again I'm pretty sure that we only create this
decomposition once, so this check is probably not too expensive.
However, the error we emit is not going to be that much more meaningful
to a user than the PETSc error you get later. Moreover, this is truly
a developer error if we trigger the assertion, so I think an assertion
is totally appropriate

lindsayad commented:
@js-jixu I believe this should fix your errors when running field split

GiudGiud reviewed (commented):
should there be a dof_indices API that fills a vector?

seems that s what we have in the places that did not need modification

roystgnr commented:
```
 /**
   * Fills the vector \p di with the global degree of freedom indices
   * for the \p node, for one variable \p vn.
   */
  void dof_indices (const Node * const node,
                    std::vector<dof_id_type> & di,
                    const unsigned int vn) const;
```
?

It's a `DofMap` method but we seem to have that handy in the code here.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1230427/) on 3facd47 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22643/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1230440/) on 3facd47 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9c78c6</th>
      <th colspan=4>#22643 3facd4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.36%</td>
      <td>84.35%</td>
      <td>-0.01%</td>
      <td>15.38%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80674</td>
      <td>80676</td>
      <td>+2</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14958</td>
      <td>14965</td>
      <td>+7</td>
      <td>22</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22643/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22643/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22643/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 15.38% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
Use DofMap vector dof indices API

lindsayad commented:
@fdkong @dewenyushu my new assertion causes a failure in the Modules Debug recipe for the contact field split test:
```
contact/test:fieldsplit.2block-3D: Assertion `libMesh::cast_int<libMesh::numeric_index_type>(split_size_sum) == dmm->_nl->nonlinearSolver()->system().get_system_matrix().local_m()' failed
contact/test:fieldsplit.2block-3D: Local split size sum 270 and local system matrix size 162 do not match.
```
Looking at `-snes_view` I see that the IS size of the `fieldsplit_contact_` is 162 which is actually the size of the system matrix. (`fieldsplit_interior_` has IS size of 108) Does this make sense? It doesn't make sense to me... I tend to assume a split will form two sets that don't have an intersection and union to the global index set. But maybe things are different for contact?

js-jixu commented:
> @js-jixu I believe this should fix your errors when running field split

I'll update my moose after fixing the bug and try it out.👍👍👍

lindsayad committed:
Get contact split correct

lindsayad committed:
Reenable mortar contact field split test

lindsayad committed:
Turn assertion into error

I realized that it's very possible that a mistake in user input
can trigger nonmatching sizes between the split sum and the system
matrix size. So I've turned the assertion into a user-facing error.
I'm going to profile the error checking to see whether it's expensive
but I don't think it will be

lindsayad commented:
Ok the error checking is not relevant in profiling 👍

lindsayad committed:
Options that scale better on bigger problems

lindsayad committed:
Cleanup unused FSP options

moosebuild commented:
Job [Conda build (ARM Mac)](https://civet.inl.gov/job/1231955/) on b8ed2d4 : invalidated by @lindsayad

lindsayad commented:
@recuero pinging you for awareness. I'm not sure how much you've played around with field splits for contact

recuero commented:
I haven't played with it, but it has the potential to help with scalability so I am interested. I see you re-enabled the mortar test, thanks for the fixes!

GiudGiud commented on framework/src/splits/Split.C:
```diff
@@ -49,12 +49,6 @@ Split::validParams()
-  params.addParam<MooseEnum>(
-      "schur_ainv",
-      SchurAInvEnum,
-      "Type of approximation to inv(A) used when forming S = D - C inv(A) B");
```
what motivates this?

lindsayad replied:
The option is never used by PETSc. It's a red herring having the option available to users

GiudGiud commented on modules/contact/test/tests/fieldsplit/tests:
```diff
@@ -11,11 +11,10 @@
-    skip = 'Weighted mortar'
-    requirement = 'The system shall allow split preconditioning of frictional mortar contact.'
+    requirement = 'The system shall allow split preconditioning of frictionless mortar contact.'
+    petsc_version = '>= 3.16.6' # early versions of petsc can't handle change in sparsity pattern in field split
```
in my limited understand of contact, frictionless and frictional would be different.
can you document why you are making that change and is it necessary for you to? does frictional not make sense here?

GiudGiud replied:
or does it just not matter

lindsayad replied:
I guess I did it because the node-face version was frictionless

lindsayad replied:
I can work on changing it back if you'd prefer

lindsayad replied:
Or I think I would add a frictional test and keep what is now the frictionless test

lindsayad replied:
I didn't think much about changing the requirement given that the test was being skipped coming into this

GiudGiud replied:
Well depends on if it matters wrt to the subject at hand

lindsayad replied:
With mortar contact you don't actually need to use the `ContactSplit` object nor the contact data members on `DM_Moose`. You can just rely on splitting by variables, whereas for node-face you don't have Lagrange multipliers so you need to split based on displacement interior and contact interface dofs, and that *does* use the `ContactSplit` object and the contact data members on `DM_Moose`. So in terms of using the contact lines of code I've touched in this PR, only the node-face test does that.

lindsayad replied:
It doesn't cost me much to add a frictional test. I'll just do that now

recuero replied:
Didn't look at the code... Just wanted to add that one may want to use field split with penalty mortar contact, without LMs.

lindsayad replied:
For which we still have the ContactSplit which we test in the node face test. 

We don’t have penalty mortar mechanical contact do we?

recuero replied:
We should have penalty mortar contact soon. 

I assume using the condensed LMs approach would make the same usage of field split as the node face test you pointed out too.

lindsayad committed:
Add frictional test and improve options

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1233851/) on 995ae70 : invalidated by @lindsayad

lindsayad commented:
Are we good to go here?

GiudGiud commented:
I think we ll want a test with vector variables and a test with array variables

bwspenc commented:
@lindsayad I'm glad to see you're working on this! What would the implications of over-providing dofs been for contact? It was a long time ago when I tried this for contact, but what I remember was that it had good convergence rates, but it was just really slow.

lindsayad commented:
I think what would happen is that you would precondition the overlapping indices twice ... and essentially whatever preconditioner went last would "win".

lindsayad committed:
Create vector var field split test

lindsayad committed:
Add array variable test

lindsayad committed:
Don't run debug on long contact tests

lindsayad commented:
Ok @GiudGiud I've added FSP framework level tests for both vector and array variables

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1253045/) on 2dc40d1 : invalidated by @lindsayad

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1253045/) on 2dc40d1 : invalidated by @lindsayad

lindsayad commented:
@GiudGiud or @roystgnr when you have time, I'd appreciate a review on this. @js-jixu has been poking me on this for a bit now and I'd like to help

GiudGiud commented:
will do tomorrow

GiudGiud commented:
This is a net and big fix so I m recommending we merge it. 
While testing it I fed it a lot of bad syntax that we should be able to catch. And I found an issue with array variable Schur field splits. 

Here we have 3 variables, u_vec, v_vec, w
**Duplicate variable specifications that matches the size**
```[Preconditioning]
  [FSP]
    type = FSP
    topsplit = 'uv'
    [uv]
      splitting = 'u v w'
      # Generally speaking, there are four types of splitting we could choose
      # <additive,multiplicative,symmetric_multiplicative,schur>
      splitting_type  = additive
    []
    [u]
      vars = 'w'
      petsc_options_iname = '-pc_type -ksp_type'
      petsc_options_value = '     hypre preonly'
    []
    [v]
      vars = 'u_vec w'
      petsc_options_iname = '-pc_type -ksp_type'
      petsc_options_value = '     hypre  preonly'
    []
    [w]
      vars = 'w'
      petsc_options_iname = '-pc_type -ksp_type'
      petsc_options_value = '     hypre  preonly'
    []
  []
[]
```

**Schur dies unless we pass these options**
```
petsc_options_iname = '-pc_fieldsplit_schur_precondition'
petsc_options_value = 'selfp'
```
with
```
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: No support for this operation for this object type
[0]PETSC ERROR: Mat type schurcomplement
[0]PETSC ERROR: See https://petsc.org/release/faq/ for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.16.6, unknown 
```

I added x_arr an array variable
**Array var Schur field splits involving dirichlet BCs (on the array var or on the other variables) break** 
See attached input
[vector-test.txt](https://github.com/idaholab/moose/files/10147011/vector-test.txt)
This could be fixed in this PR or a follow up

and unrelated
**array diffusion doesnt check the size of the arrays without an assert**
```
Assertion `(*_d_array)[_qp].size() == _var.count()' failed
diffusion_coefficient size is inconsistent with the number of components of array variable
at /Users/giudgl/projects/moose_v5/framework/src/kernels/ArrayDiffusion.C, line 53
```

lindsayad commented:
Thanks for the investigation and review. Do you want to put those into issues? Otherwise I can

GiudGiud commented:
I'll do it I have the inputs handy that show each
